### ISSUE TITLE
Feature request: Review of Recurrences

### ISSUE THREAD

**Issue by jojo12** [jemproject/JEM-Project#1762]
There was a post who said that in our JEM documentation is written, that you can repeat events once a year. But (until now) this is not possible.
Could this be a new feature?

**Comment by mckillo** [comment]
![image](https://github.com/jemproject/JEM-Project/assets/2765316/30807bc0-2b6d-4485-956c-8ad3c611d0e2)

Draft

**Comment by mckillo** [comment]
I think that the recurrence process needs a revision.
I could upload a new branch about recurrence (improvement & review setting).


For example, in the next commit (see below, test it, but allow me some minutes to finish):
- Add year option.
- When edited, show the options saved in the fields.
- New shield by days, week, month, year in the global setting.

The new branch could include an internal review of the recurrence process: 
- Create all events for all type (now there are errors because the shield is a limitation, don't create all events of series.
- When  edit and changing options, change the all events of series for type, repeat, and end of time.
- Update  the content of all events (inherited) if recurrence options and dates don't change.

 What do you think?

**Comment by mckillo** [comment]
The current shield is 60 days for all recurrence types, and all recurrence events with date > now+ 60 days aren't created in the events table.

I have to add a extension of shield for each type of recurrence in the JEM Setting -> Global Parameters ->Recurrence Events:

![image](https://github.com/jemproject/JEM-Project/assets/2765316/a3a40082-b3cd-4db0-98e4-5d6074399d69)

So, each one has the number of months for shield. Shield is a limit date because JEM should not create events ad infinity  and beyond. The admin can change the default values for each shield.

**Comment by mckillo** [comment]
![image](https://github.com/jemproject/JEM-Project/assets/2765316/05af5caa-7e3f-44ff-95d8-ba6979f748f8)

Update the commit with new icon to edit the root event of serie (only for recurrence events).

**Comment by mckillo** [comment]
Now, when the recurrence event is edited, JEM shows a mesage:
```
This event is the first one of a set of recurring events. If you save this event, the set will be dissolved, even if you just save it without editing at all.
If you choose any recurrence in this event, a new set of recurring events will be created.
```
It always happens when saving the event.

The guestions, always? Imagine.. 
If we have a recurring event and we edit the first event (root), we do not need to delete all the recurring events in the series (dissolved), in the case of changing data that does not affect the recurrence (type, number, end date), it is possible to update all recurring events with the changed data, as long as the events continue to inherit from the root event.

For example, if the 
- title
- category
- start time
- end time
- venue
- registration (in this case, depend of type registration and change).
- Publishing tab (all data)
- Other (all data)
- and Event description
is/are changed. 

It's better to update the events in the recurring series rather than delete them to create updated ones, for one main reason: the generated IDs, especially if they are shared via email and social networks.

When an event in the series that is not the first one (non-root) is edited, that event is saved as updated without the root, becoming independent. Therefore, if changes are made to the root event, it will no longer be updated as it has lost the inheritance from the series.

**Comment by mckillo** [comment]
This is a proposal to improve the editing of recurring events and maintain the inheritance of many data points of the event in the generated series.

**Comment by mckillo** [comment]
With this last commit, the recurrence is generated correctly, but the editing of the series needs review too and others improvements. To achieve this, the first element (root event) must be edited.

If any of the following data are changed in the first event:

start date
end date
recurrence type (daily, weekly, monthly, weekday, yearly)
repeating every (number)
recurrence counter (until date)
The root event will be updated, the entire previous series will be deleted, and a new one will be generated based on the saved data.

What do you think about the proposed revision of the recurrence? Is it consistent with its use?

**Comment by mckillo** [comment]
@jojo12 Maybe we need to change the title of the issue 
This is going to be broader than just adding a birthday :-)

**Comment by jojo12** [comment]
I don't understand all.
Meanwhile I was used to proced as follows: I start creating an event, then I review it, if it is complete or to correct it, then I open again and choose recurrence, then I trigger it, Then I must have the possibility to edit a single following event (because there may be single changes.
There is an other mentionned problem: Is the eventregistration for the whole recurrence, or only for each event separately?

**Comment by mckillo** [comment]
"There is an other mentionned problem: Is the eventregistration for the whole recurrence, or only for each event separately?

Good point, I have to thinnk how to do it. I'll make a note of that.

**Comment by mckillo** [comment]
I have updated the last commit to add the new edit icon in the legacy layout.

**Comment by mckillo** [comment]
An example regarding your question. I create an event for a course that will be held on the next 4 Fridays, but the registration in this case is unified, you only have to register once for the course, and it reserves a spot for all 4 Fridays. Is it right?

**Comment by jojo12** [comment]
i tested locally and got following error:
If in admin new event:
0 Failed to parse time string (now + month) at position 4 (+): Unexpected character 

DateTime->__construct() | JROOT\libraries\src\Date\Date.php:126
-- | --
3 | Joomla\CMS\Date\Date->__construct() | JROOT\administrator\components\com_jem\views\event\tmpl\edit.php:394
4 | include() | JROOT\libraries\src\MVC\View\HtmlView.php:416
5 | Joomla\CMS\MVC\View\HtmlView->loadTemplate() | JROOT\libraries\src\MVC\View\HtmlView.php:204
6 | Joomla\CMS\MVC\View\HtmlView->display() | JROOT\administrator\components\com_jem\classes\admin.view.class.php:29
7 | JemAdminView->display() | JROOT\administrator\components\com_jem\views\event\view.html.php:81
8 | JemViewEvent->display() | JROOT\libraries\src\MVC\Controller\BaseController.php:697
9 | Joomla\CMS\MVC\Controller\BaseController->display() | JROOT\administrator\components\com_jem\controller.php:51
10 | JemController->display() | JROOT\libraries\src\MVC\Controller\BaseController.php:730
11 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT\administrator\components\com_jem\jem.php:54
12 | require_once() | JROOT\libraries\src\Dispatcher\LegacyComponentDispatcher.php:71
13 | Joomla\CMS\Dispatcher\LegacyComponentDispatcher::Joomla\CMS\Dispatcher\{closure}() | JROOT\libraries\src\Dispatcher\LegacyComponentDispatcher.php:73
14 | Joomla\CMS\Dispatcher\LegacyComponentDispatcher->dispatch() | JROOT\libraries\src\Component\ComponentHelper.php:361
15 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT\libraries\src\Application\AdministratorApplication.php:150
16 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT\libraries\src\Application\AdministratorApplication.php:195
17 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT\libraries\src\Application\CMSApplication.php:306
18 | Joomla\CMS\Application\CMSApplication->execute() | JROOT\administrator\includes\app.php:58
19 | require_once() | JROOT\administrator\index.php:32

**Comment by mckillo** [comment]
@jojo12 What number have you in SJEM Setting -> Global Parameters ->Recurrence Events:? By default will be ->
![image](https://github.com/jemproject/JEM-Project/assets/2765316/754dedba-6519-4830-b617-71cc16f9b86b)

I think that these fields is empty or don't create, if it's so, you need to update the _jem_config table with sql file 
admin/sql/updates/mysql/4.2.3.sql
this file is in the commit.

**Comment by jojo12** [comment]
Sorry, Marcos, I have a cold and I am after all in bed. 
I installed your review as package and I can create and edit events. Will continue to test when I stay better.

**Comment by mckillo** [comment]
No problems, no rush, you have to recover first.

**Comment by mckillo** [comment]
ALPHA 2 version: New option 'bookedrootevent" in the event table. This field is for registration setting, if it's set (value =1), indicates that the booking places is in the root event of series and for all events of serie (recurrence events with that root ID). 
In this commit, this field is saved in the table for all events, and it can be edited.
Coming soon, beta 1 version with registration process users.

**Comment by mckillo** [comment]
If you apply the last commit, remember to run the sql files.

**Comment by jojo12** [comment]
I zipped in the branch the com_jem and installed. But then in admin I can create events, but I can't see them in frontend.
But the setting look like you describe, I will retry later.

**Comment by jojo12** [comment]
when I install the zipped com_jem from recurrence_review, I get
**Warnung
JInstaller: :Install: SQL-Fehler Unknown column 'ujrfe_jem_config.keyname' in 'where clause'**

**Comment by mckillo** [comment]
> @jojo12 What number have you in JEM Setting -> Global Parameters ->Recurrence Events:? By default, they will be ->
> ![image](https://github.com/jemproject/JEM-Project/assets/2765316/754dedba-6519-4830-b617-71cc16f9b86b)
> 
> Check them. I think that these fields are empty or don't create, if it's so, you need to update the _jem_config table with SQL file 
> admin/sql/updates/mysql/4.2.3.sql
> This SQL file is in the commit, run to mysql.

@jojo12  I think that you don't have aplied 4.2.3.sql file (run into your mysql).

**Comment by jojo12** [comment]
what should I do with the 4.2.3.sql the first line with this error?

**Comment by jojo12** [comment]
first I was not able to enter,
Then I reinstalled the com_jem.zip with the strange line in 4.2.3.sql
Then I was able to enter, but I wasn't able to check the "booked in a root event" after save, reopened and there was no check.

**Comment by jojo12** [comment]
I have these numbers for recurrence.

**Comment by mckillo** [comment]
You are right, the line 2 of 4.2.3.sql had the sufix joomla. I already solve it.
Now, you can run this file in mysql by phpmyadmin. Check in your _jem_config table these fields:

![image](https://github.com/jemproject/JEM-Project/assets/2765316/447d3875-17f9-4ec8-ab52-a5857bffcf8c)

**Comment by mckillo** [comment]
The values in image capture are test values (by default are 3,60,12,180)

**Comment by mckillo** [comment]
I have testing in other languages, I will try to simulate the error, but the code doesn't read these numbers. Send me a capture of your event (recurrence and registration setting).

**Comment by jojo12** [comment]
in frontend event detail I get:
`Warning: Undefined property: JemViewEvent::$1 in C:\xampp\htdocs\joomla\components\com_jem\views\event\tmpl\responsive\default_attendees.php on line 63`

**Comment by mckillo** [comment]
@jojo12 You have a new commit to solve it. Sorry, the call variable bookedrootevent was wrong in this view.

**Comment by mckillo** [comment]
I have to upload the code of beta 1 with registration in the root event, but I have to finish the tests. I hope to do it this weekend.

**Comment by mckillo** [comment]
In the design of new recurrence event, I have created a new option “How booked in the root event  when the event is recurrence”. This new field in the event table is 'bookedrootevent' as INT (4 Bytes: 00 00 00 00 = Not used - Max - Min - Type).

PROPOSAL to extender use
bookedrootevent can be 3 values (type):
0: No active, each event has its registration.
1: The registration is in all recurrence events.
2. The registration is only in the selected recurrence events (this type can be limited to min & max selected events).

I'll try to dev this model in the beta branch.

For example, the admin can create a recurrence event with 10 recurrences in the time, and set bookedrootevent of type=2 with min=1 and max=2 selected events. The user can see all 10 events and in the registration form, can select in what recurrence event wants to book places with a min=1 and max=2.

**Comment by webenhanced** [comment]
Hi folks, while you are reviewing the recurrence system, I have a request.

When using Weekday recurrence, there is a drop down for "repeating every" that has a list including 1-4, last, before last.

I have a need to repeat an event on the 1st, 3rd, and 5th Sundays of each month. The 5th Sunday only happens a few times during the year, and when it does the event runs two Sundays in a row. So most months there are two events and sometimes there are three events in the month.

"last" does not work for this situation.

Please add 5th weekday to the available dropdown list.

**Comment by mckillo** [comment]
@webenhanced Here you are, in the last commit (see on the line above) is added your request. Check it and tell me if it works for you.

**Comment by jojo12** [comment]
I installed the nightly for recurrence review (hekla did) on a localhost:
when I choose "weekday" it should open a submenu, but nothing! This is a bug!

**Comment by jojo12** [comment]
and the information in backend is bad:
![grafik](https://github.com/jemproject/JEM-Project/assets/2960236/efda9c62-a0c8-454b-b905-a6ed42624442)

**Comment by mckillo** [comment]
@jojo2 Try it with the last commit (beta 1)

![image](https://github.com/jemproject/JEM-Project/assets/2765316/6a9a502e-cfc6-4e75-bb4b-4a3245f76abd)

![image](https://github.com/jemproject/JEM-Project/assets/2765316/a8f50fe3-66c6-4269-a5a6-6d5ff333f380)

![image](https://github.com/jemproject/JEM-Project/assets/2765316/2f52eb4a-c75c-4f2c-b1c6-c651ef357afb)

![image](https://github.com/jemproject/JEM-Project/assets/2765316/039eb794-7cf9-47c6-ad49-36905cfdb970)

I have to do more test some combination of settings. Your advice is appreciated to improvement this feature.

**Comment by mckillo** [comment]
Run the sql file, there is changes.

**Comment by jojo12** [comment]
Now I don't get any event in Frontend simplelist!
the rest seems good

**Comment by mckillo** [comment]
I have reviewed my FE simplelist and It works (I tested it changes languages, in EN and in ES). I don't know why you don't see the event in this view. Tell me How and what version do you have installed JEM and branch?

**Comment by mckillo** [comment]
@jojo12 I have updated the beta 1 (see last commit)
Fix small bug.
Update addScriptDeclaration to addInlineScript in the code (support J5 & J6).
Improve style in event view (FE) for some select (hours & minutos) and input field (venues).

**Comment by jojo12** [comment]
? Where is the actual one?

**Comment by mckillo** [comment]
In the root of branch https://github.com/jemproject/JEM-Project/tree/recurrence-review you have the last zip file by Helka

**Comment by jojo12** [comment]
todo: add in jem_mailer plugin

**Comment by jojo12** [comment]
todo: Backend Reg.Users list: should have a column for "root event booking"

**Comment by webenhanced** [comment]
I also don't get any events in Simplelist view.
I first installed pkg_jem_nightly_build_20240706-v1.zip and created weekday repeating event. It generated the correct number of repeats in the admin interface (and I could see them in the DB). No front end events shown.
I then noticed a new nighly (pkg_jem_nightly_build_20240707-v1.zip) and installed that over the top of the above install. Install went perfect.

I then created a bi-weekly recurring event - still nothing on frontend.
Then created a single day entry and still not shown on frontend.
Menu setting for FE are:
- archive events = yes
- days before = 15
- days after = 60
I notice that the Published column in the event table was "-2" for most of the entries, so don't know if that means anything.

Is there anything you'd like me to try? Let me know.

**Comment by mckillo** [comment]
I tried it and it works well for me. I can't simulate this error. I have installed a new JL 5.1.1 on PHP8.2 with JEM (pkg_jem_nightly_build_20240706-v1.zip, the same version as you)
If the Published column is -2, the events have been deleted, this is very rare.

I send the event creation and the setting of the eventlist menu.

![image](https://github.com/jemproject/JEM-Project/assets/2765316/59500766-bdcb-4a83-8e4c-102b82bad702)

![image](https://github.com/jemproject/JEM-Project/assets/2765316/b3c3df6b-1c13-4c6d-9655-abaacbf86331)

![image](https://github.com/jemproject/JEM-Project/assets/2765316/5042987e-c7eb-4ef0-be08-81c77df49b53)

Can you send a capture of events and setting menu?

**Comment by mckillo** [comment]
I think that this issue comes from JEM update. Tell me if you have any JEM version installed before to install pkg_jem_nightly_build_20240706-v1.zip
The installation script doesn't have modified because this version is a beta (It's better to test as a new JEM).

**Comment by webenhanced** [comment]
I did have JEM 4.2.2 on there originally, then when installing over the top, I got an error with DB - one of the occurrence_ fields was already there.

So then I uninstalled JEM and then installed pkg_jem_nightly_build_20240706-v1.zip and the issue was there.

I'll clean out the jem db tables after uninstalling it and start from fresh at 20240706-v1.

**Comment by mckillo** [comment]
OK, you have a new zip https://github.com/jemproject/JEM-Project/blob/recurrence-review/pkg_jem_nightly_build_20240707-v1.zip, and you must apply the last commit of the morning: 9c836a7ab0404a18cefb8f325776c1a6b4d67304

Thanks for info of your updete, I'll try to do it for simulate the issue.

**Comment by webenhanced** [comment]
Hi, sorry I've been absent for a while. I had to build a better dev environment here on my LAN. Got that happening OK now.
I have tested the pkg_jem_nightly_build_20240725-v1.zip from the 4.2.3 dev branch after removing the previous version, and all jem* tables from the DB.
Install went well; created 1st sunday of month - OK; created 5th sunday of month and for this month it created an entry for the 18th. When viewing that entry, in the recurrence box it says:-

> weekday, repeating every  last.  per month
Has this got anything to do with what date is chosen in the event creation page? Is the software treating 5th as "last"?
Other than that, it seems to be going well.
Anything you'd like me to try out?
